feat(ux/ui): implement toaster,support noIcon property, error toast dismissal, mobile on close

cynthiaricomendoza-alaskaair opened the pull request:
# Alaska Airlines Pull Request

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

**Fixes:** #19, #9, #22, #26

## Summary:

Created a toaster container to hold unlimited number of toasts.
Added noIcon property to allow the icon to be remove from toast UI.
Error toast must be dismissed by user - do not apply auto dismissal logic to error toasts.
For screen sizes < 767, allow the user to dismiss the toast by clicking anywhere on the toast.


## Type of change:
- [x] New capability

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

cynthiaricomendoza-alaskaair committed:
feat(ux/ui): implement toaster #19

This commit completely implement the toaster.js feature
that will manage the experience of multiple toasts.

Changes to be committed:
modified:   README.md
modified:   apiExamples/basic.html
modified:   apiExamples/error.html
new file:   apiExamples/multipleToasts.html
modified:   apiExamples/success.html
renamed:    demo/toggleToast.js -> demo/addToast.js
modified:   demo/apiExamples.md
modified:   demo/demo.md
modified:   demo/index.html
modified:   docs/api.md
modified:   docs/partials/demo.md
modified:   docs/partials/description.md
modified:   docs/partials/useCases.md
modified:   package-lock.json
modified:   package.json
modified:   src/auro-toast.js
new file:   src/auro-toaster.js
new file:   src/style-toaster.scss
modified:   src/style.scss
modified:   test/auro-toast.test.js

cynthiaricomendoza-alaskaair committed:
feat(ux/ui): support noIcon property #9

DZwell commented on src/auro-toaster.js:
```diff
@@ -0,0 +1,69 @@
+
+  }
+
+  connectedCallback() {
```
Do we need to make a call to `disconnectedCallback` to clean up the event listeners?

cynthiaricomendoza-alaskaair replied:
yes i believe we do! thanks for catching that. i pushed up some changes for this

DZwell commented on src/auro-toast.js:
```diff
@@ -64,12 +71,26 @@ export class AuroToast extends LitElement {
    */
   handleOnClose() {
-    this.visible = false;
+    this.fadeOutToast();
```
Would we want to call `closeToast` here as well? I would think we don't really care about fading out the toast when a user clicks the X, we just want it to disappear.

cynthiaricomendoza-alaskaair replied:
i think we still want to fade it out and close it, so that's why i added `closeToast()` the timeout in `fadeOutToast()`

blackfalcon commented on apiExamples/noIcons.html:
```diff
@@ -0,0 +1,3 @@
+<auro-button onClick="addToast({message: 'Hi, I am a default toast!', noIcon: true})">Show default toast</auro-button>
+<auro-button onClick="addToast({variant: 'error', message: 'Hi, I am an error toast!', noIcon: true})">Show error toast</auro-button>
+<auro-button onClick="addToast({variant: 'success', message: 'Hi, I am a success toast!', noIcon: true})">Show success toast</auro-button>
```
Please see editorconfig 

```
insert_final_newline = true
```

blackfalcon commented on demo/addToast.js:
```diff
@@ -3,12 +3,9 @@
+    detail: toastDetails
+  });
+  document.dispatchEvent(addToastEvent);
+}
```
Add line break, please see editor config.

cynthiaricomendoza-alaskaair committed:
feat: do not auto-dismiss error toast #26

cynthiaricomendoza-alaskaair committed:
perf(mobile): allow greater touch surface #22

This commit will allow users to tab anywhere inside
the toast when dismissing.

Changes to be committed:
modified:   src/auro-toast.js
modified:   src/style-toaster.scss
modified:   src/style.scss

jessieiswise commented on src/style-toaster.scss:
```diff
@@ -0,0 +1,48 @@
+@import './../node_modules/@aurodesignsystem/webcorestylesheets/dist/utilityClasses/displayProperties';
+
+// Support for auroElement styles
+// @import './node_modules/@aurodesignsystem/webcorestylesheets/dist/auroElement/auroElement';
```
should this be commented out?

cynthiaricomendoza-alaskaair replied:
i _think_ it's ok commented out for now. it's commented out because that's how the template of this repo had it. @blackfalcon, is this still ok to be commented out?

blackfalcon replied:
Either leave the comment as an instructional reference or delete the code if not used.

jessieiswise commented on src/style-toaster.scss:
```diff
@@ -0,0 +1,48 @@
+  pointer-events: none;
+  
+  @media ($auro-grid-breakpoint-sm <=width < $auro-grid-breakpoint-lg) {
+    right: 1%;
```
this is set in both breakpoints. may just want to make one query to cover both use cases

cynthiaricomendoza-alaskaair replied:
ahh yes you're right. this might've been left over from a merge conflict i resolve, thanks for catching that!

jessieiswise commented on src/style-toaster.scss:
```diff
@@ -0,0 +1,48 @@
+}
+
+// Selector for native :focus-visible support
+:focus-visible {
```
what element is this selector for? I'd feel more comfortable with a less vague selector. Encapsulation makes it not really a big deal but could trip up future work in this component

blackfalcon replied:
This selector should NOT be used. This was for a [polyfill](https://github.com/WICG/focus-visible) we no longer support and is not installed in this project. 

Please refer to the current [pseudo-class](https://css-tricks.com/almanac/selectors/f/focus-visible/) spec.

jessieiswise commented on src/auro-toast.js:
```diff
@@ -81,30 +103,79 @@ export class AuroToast extends LitElement {
+
+
+  updated(changedProperties) {
+    this.setOnClick();
```
`setOnClick` will run every time the `updated` lifecycle method runs- would `connectedCallback` be a more appropriate lifecycle method to invoke this?

cynthiaricomendoza-alaskaair replied:
i like that suggestion, i'll make sure to push some changes for this

jessieiswise reviewed (approved):
Great job! I had a few comments but nothing blocking, consider them an fyi. LGTM!

blackfalcon commented on README.md:
```diff
@@ -21,7 +21,7 @@ The following sections are editable by making changes to the following files:
 <!-- AURO-GENERATED-CONTENT:START (FILE:src=./../docs/partials/description.md) -->
 <!-- The below content is automatically added from ./../docs/partials/description.md -->
-`<auro-toast>` is a [HTML custom element](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) for the purpose of illustrating toast messages to Alaska customers.
+`<auro-toaster>` is a [HTML custom element](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) for the purpose of illustrating a stack of toast messages to Alaska customers.
```
The README is not clearly defining the difference between a toast and the toaster.

blackfalcon commented on apiExamples/basic.html:
```diff
@@ -1,2 +1,3 @@
-<auro-toast>Toast content goes here.</auro-toast>
+<auro-button onClick="showToast('#defaultToast')">Show default toast</auro-button>
+<br>
+<auro-toast id="defaultToast"> Default toast with no error type  </auro-toast>
```
Please see editorconfig 

```
insert_final_newline = true
```

blackfalcon commented on apiExamples/error.html:
```diff
@@ -1 +1,3 @@
\ No newline at end of file
+<auro-button onClick="showToast('#errorToast')">Show error toast</auro-button>
+<br>
+<auro-toast variant="error" id="errorToast"> Unable to add lap infant. Please try again  </auro-toast>
```
Please see editorconfig 

```
insert_final_newline = true
```

blackfalcon commented on apiExamples/multipleToasts.html:
```diff
@@ -0,0 +1,10 @@
+    <auro-toast id="toast-default">Default toast</auro-toast>
+    <auro-toast id="toast-error" variant="error">Unable to add lap infant. Please try again</auro-toast>
+    <auro-toast id="toast-success" variant="success">Successfully added lap infant</auro-toast>
+</auro-toaster>
```
Please see editorconfig 

```
insert_final_newline = true
```

blackfalcon commented on apiExamples/success.html:
```diff
@@ -1 +1,3 @@
\ No newline at end of file
+<auro-button onClick="showToast('#successToast')">Show success toast</auro-button>
+<br>
+<auro-toast variant="success" id="successToast"> Successfully added lap infant  </auro-toast>
```
Please see editorconfig 

```
insert_final_newline = true
```

blackfalcon commented on demo/addToast.js:
```diff
@@ -3,12 +3,10 @@
+    toast.setAttribute('visible', true);
   }
-};
+};
```
Please see editorconfig 

```
insert_final_newline = true
```

blackfalcon commented on demo/apiExamples.md:
```diff
@@ -7,20 +7,23 @@ The auro-toast element provides users a way to ... (it would be great if you fil
-| [type](#type)    | `String`  | Component will render visually based on which type value is set; currently supports `error`, `success` |
+| Attribute | Type      | Description                               |
+|-----------|-----------|-------------------------------------------|
+| [fixed](#fixed)   | `Boolean` | Uses fixed pixel values for element shape |
```
This attr is no longer supported. Please remove the following line from ./src/auro-toast.js

```
@attr {Boolean} fixed - Uses fixed pixel values for element shape
```

blackfalcon commented on apiExamples/error.html:
```diff
@@ -1 +1,3 @@
-<auro-toast variant="error">Toast content goes here.</auro-toast>
\ No newline at end of file
+<auro-button onClick="showToast('#errorToast')">Show error toast</auro-button>
+<br>
```
Not a fan of using HTML as layout. I'd rather see inline CSS to create space if needed versus promoting an unwanted pattern. 

Please update all these partial files.

blackfalcon commented on src/auro-toast.js:
```diff
@@ -81,30 +103,82 @@ export class AuroToast extends LitElement {
+          <button class="closeButton" aria-label="closeToast" @click="${this.handleOnClose}">
+            ${this.svg}
+          </button>
+      </div>`
```
Intentations are indorrect. Line 177 opens and closes a DIV. `<button>` is a sibling, not a child. 

```suggestion
    return this.visible ? html`<div id="toastContainer">
      ${iconHtml}
        <div class="message"><slot></slot></div>
        <button class="closeButton" aria-label="closeToast" @click="${this.handleOnClose}">
          ${this.svg}
        </button>
    </div>`
```

jessieiswise replied:
@blackFalcon, what's your take on adding an auto linting commit hook?

blackfalcon replied:
That's a debate at the Generator level. If you have ideas, submit them there.

blackfalcon commented on src/style-toaster.scss:
```diff
@@ -0,0 +1,43 @@
+  background-color: var(--auro-color-border-error-on-light);
+  color: var(--auro-color-base-white);
+
+}
```
See editor config, add line break.

blackfalcon commented on src/style.scss:
```diff
@@ -14,70 +14,79 @@
   display: inline-flex;
-  min-width: 95%;
+  min-width: 100%;
   min-height: 60px;
```
I rally want to create `var(--auro-size-750)` for this value.

cynthiaricomendoza-alaskaair replied:
would you like me to create an issue in the Auro Design repo and create a value for that? or should i add a comment on here to update this value once that is done?

cynthiaricomendoza-alaskaair replied:
@blackfalcon this is the response i got from designers about the min-width being 60px, "let’s keep it at 60px for now… we’ll get back to you Friday at the latest if we want to make a change if that’s okay?" designers are out at a design summit today and tomorrow. 

i'm kind of in a pickle here; i'm debating on going ahead and creating a new token for this or wait till friday to hear back from them? i would like to create one to keep things going, but i know if they end up wanting to use a pixel size that already exists as a token we won't need that token anymore. what are you thoughts?

blackfalcon replied:
We're good. Not a critical step.

blackfalcon commented on src/style.scss:
```diff
@@ -14,70 +14,79 @@
+    display: flex;
+    justify-content: center;
+    padding-top: var(--auro-size-25);
+    margin-right: -8px;
```
you can use negative values with custom properties, `var(--auro-size-100)`

blackfalcon commented on test/auro-toast.test.js:
```diff
@@ -1,4 +1,5 @@
-import { fixture, html, expect, aTimeout } from '@open-wc/testing';
+import { fixture, html, expect, aTimeout, elementUpdated } from '@open-wc/testing';
+import sinon from 'sinon'
```
I don't think sinon is being used.

cynthiaricomendoza-alaskaair committed:
chore(deps): remove unused dependency

this commit will remove the dependency @open-wc/testing-karma, since we don't use it anymore

Changes to be committed:
modified:   package-lock.json
modified:   package.json

blackfalcon commented on demo/apiExamples.md:
```diff
@@ -1,169 +1,190 @@
+
+### Method Examples
+
+#### <a name="methodName"></a>`methodName`<a href="#auro-toast" style="float: right; font-size: 1rem; font-weight: 100;">back to top</a>
```
Please review this document, I am not sure this is doing what we may expect it to do here.

blackfalcon commented on demo/demo.md:
````diff
@@ -2,52 +2,187 @@
+  <auro-toaster>
+```
+
 ## Recommended Use and Version Control
````
The demo here is not working as expected.

cynthiaricomendoza-alaskaair replied:
i pushed up changes to fix this on this commit: https://github.com/AlaskaAirlines/auro-toast/commit/c96cf10f03524f9f954ad160ef09d581e4fa806e#diff-b18ccd15ec95e2a9c4fead5bca9085042af688361a24d3463c27138e3c797a5b

blackfalcon commented on src/style.scss:
```diff
@@ -11,73 +11,79 @@
     max-width: 500px;
   }
 
+  #toastContainer {
```
We typically do not use ID as CSS selectors. https://auro.alaskaair.com/css/conventions

blackfalcon commented on src/style.scss:
```diff
@@ -11,73 +11,79 @@
+
+  }
+
+  button {
```
You already have the class of `closeButton` on the DOM, so that should be used.

cynthiaricomendoza-alaskaair committed:
docs(api): add property examples

this commit adds indiviual examples for the properties the toast supports

Changes to be committed:
new file:   apiExamples/noIcon.html
new file:   apiExamples/variant.html
new file:   apiExamples/visible.html
modified:   demo/apiExamples.md
modified:   docs/partials/apiExamples.md

cynthiaricomendoza-alaskaair committed:
docs(readme): cleanup dynamic toasts & custom element

this commit adds some code to better the example for the dynamic toasts section and cleans up the custom element section so it works and shows correctly

Changes to be committed:
modified:   demo/demo.md
modified:   demo/index.html
modified:   docs/partials/demo.md

cynthiaricomendoza-alaskaair committed:
refactor(style): cleanup toast classes

this commit takes care of the following:
removes the usage of an ID on a element in the toast and replaces it with a class
removes styles for button element since there is a class that wraps around the button and can take care of the styles there

Changes to be committed:
modified:   src/auro-toast.js
modified:   src/style.scss
modified:   test/auro-toast.test.js

cynthiaricomendoza-alaskaair committed:
perf(demo): create api.html file

this commit creates the api.html file to help review the apiExmaples.md file

Changes to be committed:
new file:   demo/api.html

cynthiaricomendoza-alaskaair committed:
perf(screenreader): notify when toast is visible

this commit adds aria-live="polite" to the auro-toast, so the screen reader captures the toast and reads that it is visible

Changes to be committed:
modified:   src/auro-toast.js

blackfalcon commented on demo/api.html:
```diff
@@ -0,0 +1,45 @@
+    <main></main>
+
+    <script type="module">
+      import 'https://unpkg.com/marked@latest/marked.min.js';
```
```suggestion
      import 'https://cdn.jsdelivr.net/npm/marked@latest/marked.min.js';
```

blackfalcon commented on demo/api.html:
```diff
@@ -0,0 +1,45 @@
+      type="text/css"
+      href="https://unpkg.com/prismjs@1.24.1/themes/prism.css"
+    />
+    <link rel="stylesheet" href="https://unpkg.com/@alaskaairux/design-tokens@latest/dist/tokens/CSSCustomProperties.css">
```
```suggestion
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@alaskaairux/design-tokens@latest/dist/tokens/CSSCustomProperties.css">
```

blackfalcon commented on demo/api.html:
```diff
@@ -0,0 +1,45 @@
+    <link
+      rel="stylesheet"
+      type="text/css"
+      href="https://unpkg.com/prismjs@1.24.1/themes/prism.css"
```
```suggestion
      href="https://cdn.jsdelivr.net/npm/prismjs@1.24.1/themes/prism.css"
```

blackfalcon commented on demo/api.html:
```diff
@@ -0,0 +1,45 @@
+      href="https://unpkg.com/prismjs@1.24.1/themes/prism.css"
+    />
+    <link rel="stylesheet" href="https://unpkg.com/@alaskaairux/design-tokens@latest/dist/tokens/CSSCustomProperties.css">
+    <link rel="stylesheet" type="text/css" href="https://unpkg.com/@alaskaairux/webcorestylesheets@latest/dist/demoWrapper.css" />
```
```suggestion
    <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/@alaskaairux/webcorestylesheets@latest/dist/demoWrapper.css" />
```

blackfalcon commented on demo/api.html:
```diff
@@ -0,0 +1,45 @@
+    />
+    <link rel="stylesheet" href="https://unpkg.com/@alaskaairux/design-tokens@latest/dist/tokens/CSSCustomProperties.css">
+    <link rel="stylesheet" type="text/css" href="https://unpkg.com/@alaskaairux/webcorestylesheets@latest/dist/demoWrapper.css" />
+    <link rel="stylesheet" type="text/css" href="https://unpkg.com/@alaskaairux/webcorestylesheets@latest/dist/elementDemoStyles.css" />
```
```suggestion
    <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/@alaskaairux/webcorestylesheets@latest/dist/elementDemoStyles.css" />
```

blackfalcon commented on demo/api.html:
```diff
@@ -0,0 +1,45 @@
+
+    <script type="module">
+      import 'https://unpkg.com/marked@latest/marked.min.js';
+      import 'https://unpkg.com/prismjs@latest/prism.js';
```
```suggestion
      import 'https://cdn.jsdelivr.net/npm/prismjs@latest/prism.js';
```

blackfalcon commented on demo/api.html:
```diff
@@ -0,0 +1,45 @@
+    </script>
+    <script type="module" src="../src/index.js"></script>
+
+    <script src="https://unpkg.com/@alaskaairux/auro-accordion@latest/dist/auro-accordion__bundled.js" type="module"></script>
```
```suggestion
    <script src="https://cdn.jsdelivr.net/npm/@alaskaairux/auro-accordion@latest/dist/auro-accordion__bundled.js" type="module"></script>
```

blackfalcon commented on demo/api.html:
```diff
@@ -0,0 +1,45 @@
+    <script type="module" src="../src/index.js"></script>
+
+    <script src="https://unpkg.com/@alaskaairux/auro-accordion@latest/dist/auro-accordion__bundled.js" type="module"></script>
+    <script src="https://unpkg.com/@alaskaairux/auro-icon@latest/dist/auro-icon__bundled.js" type="module"></script>
```
```suggestion
    <script src="https://cdn.jsdelivr.net/npm/@alaskaairux/auro-icon@latest/dist/auro-icon__bundled.js" type="module"></script>
```

blackfalcon commented on demo/api.html:
```diff
@@ -0,0 +1,45 @@
+
+    <script src="https://unpkg.com/@alaskaairux/auro-accordion@latest/dist/auro-accordion__bundled.js" type="module"></script>
+    <script src="https://unpkg.com/@alaskaairux/auro-icon@latest/dist/auro-icon__bundled.js" type="module"></script>
+    <script src="https://unpkg.com/@alaskaairux/auro-button@latest/dist/auro-button__bundled.js" type="module"></script>
```
```suggestion
    <script src="https://cdn.jsdelivr.net/npm/@alaskaairux/auro-button@latest/dist/auro-button__bundled.js" type="module"></script>
```

blackfalcon commented on demo/api.html:
```diff
@@ -0,0 +1,45 @@
+    <script type="module">
+      import 'https://unpkg.com/marked@latest/marked.min.js';
+      import 'https://unpkg.com/prismjs@latest/prism.js';
+      fetch('/demo/apiExamples.md')
```
Getting error 

```
showToast is not defined
```

cynthiaricomendoza-alaskaair replied:
resolved in this commit: https://github.com/AlaskaAirlines/auro-toast/pull/20/commits/ee84d306d922046648e8a9ad628b830087783023

blackfalcon commented on src/auro-toast.js:
```diff
@@ -78,33 +99,85 @@ export class AuroToast extends LitElement {
+        break;
+    }
+
+    return this.visible ? html`<div class="toastContainer">
```
```suggestion
    return this.visible ? html`<div class="toastContainer" aria-live="polite">
```

cynthiaricomendoza-alaskaair replied:
resolved in this commit: https://github.com/AlaskaAirlines/auro-toast/pull/20/commits/9815e230ab260d724f4804ae3d84368f487db059

cynthiaricomendoza-alaskaair committed:
docs(api): add missing scripts

this commit adds scripts needed to perform showing the toast when clicking on the buttons

Changes to be committed:
modified:   demo/api.html
modified:   demo/apiExamples.md
modified:   demo/index.html
modified:   docs/api.md
modified:   src/auro-toast.js

blackfalcon reviewed (changes requested):
Review demo... I don't think this is the expected result for multiple toasts. 

<img width="1139" alt="Screen Shot 2023-06-30 at 5 14 27 PM" src="https://github.com/AlaskaAirlines/auro-toast/assets/181089/1ac807cf-6d9a-4c0d-ae7f-470679a1f337">

cynthiaricomendoza-alaskaair committed:
docs(readme): add script for toaster

this commit adds the script for toaster so the mutliple toast section in the demo works correctly.

Changes to be committed:
modified:   demo/index.html

cynthiaricomendoza-alaskaair commented:
fixed! <img width="1369" alt="Screenshot 2023-07-03 at 11 49 35" src="https://github.com/AlaskaAirlines/auro-toast/assets/116121736/e86499ad-c88b-41db-84c1-72ffad69e2ab">
